2843: Remove deprecated analytics api MEL-87

kael89 opened the pull request:
### Issue #:
https://github.com/beyondessential/tupaia-backlog/issues/2843

### Changes
* Removed the deprecated api and `useDeprecatedApi` flag for event analytics
* Changed the default value of the `useDeprecatedApi` flag for events from `true` to `false`. This aligns with a lot of new code I found in the codebase, which assumes that the "modern" api is used, while not setting the flag to `false`. I guess this happens to work in practice since most new code will use `tupaia` events, which are aligned to the "modern" dhis events
* Changed the default data value format in `DhisApi.getEvents` from `array` to `object`. We use the latter in almost all cases, since it's more convinient
* Changed the schema for  `dhis` events using the "object" format, to align with the one used by modern `dhis` and `tupaia` events

A note on the schema of `event.dataValues`:
* `tupaia` events: `Record<string, string>`
* `dhis` events:
  * modern (constructed from event analytics): `Record<string, string>`
  * deprecated (constructed from events):
    * `array`: `{ dataElement, value, ... }[]`
    * `object`: `Record<string, { dataElement, value, ...}>` => _this has been changed now to also be `Record<string, string>`_


Ideas for future improvements:
* `dhis` events include some extra fields that we don't use in practice and which are not supported in `tupaia` events. Ideally, we would remove those before returning events in `DhisService` (data-broker), to ensure consistency between services

kael89 commented on packages/data-broker/src/services/dhis/pullers/EventsPuller.js:
```diff
@@ -80,7 +79,7 @@ export class EventsPuller {
 
     // TODO remove `useDeprecatedApi` option as soon as `pullEventsForApi_Deprecated()` is deleted
-    const { useDeprecatedApi = true } = options;
+    const { useDeprecatedApi = false } = options;
```
I would also love for us to delete the deprecated `dhis` event api, but we cannot currently because of this bug: https://github.com/beyondessential/tupaia-backlog/issues/818

The modern api uses event analytics under the hood, and will use `0` for empty values

IgorNadj commented on packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfEvents.js:
```diff
@@ -63,8 +52,12 @@ class TableOfEventsBuilder extends DataBuilder {
 
+  getDataElementCodes() {
+    const { dataElementCodes, columns } = this.config;
+    return dataElementCodes ?? Object.keys(columns).filter(key => !isMetadataKey(key));
```
Ah good to know. Would be useful as a comment in the code in data broker's tupaia service pull events method. And maybe throw an error explicitly if dataElementCodes is empty?

kael89 replied:
You can still fetch tupaia events with data element codes, but the `dataValues` will be empty (this is used eg when you are just counting events) - implemented in https://github.com/beyondessential/tupaia-backlog/issues/2636

IgorNadj replied:
Ah makes sense. Maybe a comment somewhere would be useful for those like me?

IgorNadj commented on packages/data-broker/src/services/dhis/DhisTranslator.js:
```diff
@@ -140,34 +133,6 @@ export class DhisTranslator {
     return this.inboundAnalyticsTranslator.translate(response, dataSources);
   };
 
-  translateInboundEventDataValues = (dataValues, dataElementToSourceCode) => {
```
Great cleanup!

IgorNadj commented on packages/data-broker/src/services/dhis/pullers/AnalyticsPuller.js:
```diff
@@ -240,13 +190,8 @@ export class AnalyticsPuller {
-      return useDeprecatedApi
-        ? this.pullAnalyticsFromEventsForApi_Deprecated
-        : this.pullAnalyticsFromEventsForApi;
+      return this.pullAnalyticsFromEventsForApi;
```
It's beautiful

IgorNadj reviewed (approved):
Massive effort, thanks for doing it!

edmofro commented:
Yeah awesome work @kael89, thanks for the spring clean!

kael89 commented:
@IgorNadj I rebased from `dev` (no conflicts). Also, I added a fix (https://github.com/beyondessential/tupaia/pull/2928/commits/b2a53224b8c8929e0c44bde02738090f91eddaf1). E2e tests for dashboard reports came out fine 😃 I still have to run tests for map overlays, waiting for their config to be built by our testers

kael89 commented:
I also updated a dashboard report whose response resulted in flaky e2e tests (https://github.com/beyondessential/tupaia/pull/2928/commits/86fb89a73047cf55753a3bfcb9ffe3b46fcd7cd0)
